### ISSUE TITLE
TypeError: fastify_plugin_1.default is not a function

### ISSUE THREAD

**Issue by persona0591** [fastify/fastify-plugin#89]
## 🐛 Bug Report

I'm trying to use `fastify-plugin` with TypeScript. However, the error underneath is thrown:

`TypeError: fastify_plugin_1.default is not a function`

Full stack:

```
TypeError: fastify_plugin_1.default is not a function
    at Object.<anonymous> (/usr/src/app/build/src/plugin.js:5:57)
    at Module._compile (internal/modules/cjs/loader.js:759:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:770:10)
    at Module.load (internal/modules/cjs/loader.js:628:32)
    at Function.Module._load (internal/modules/cjs/loader.js:555:12)
    at Module.require (internal/modules/cjs/loader.js:666:19)
    at require (internal/modules/cjs/helpers.js:16:16)
    at /usr/src/app/build/src/server.js:7:25
    at Object.<anonymous> (/usr/src/app/build/src/server.js:13:3)
    at Module._compile (internal/modules/cjs/loader.js:759:30)
```

## To Reproduce

File "src/plugin.ts" (taken from `https://github.com/fastify/fastify-plugin/blob/master/plugin.test-d.ts`)
```js
import fp from 'fastify-plugin';
import { FastifyPlugin } from 'fastify';

export interface TestOptions {
  customNumber: number
}

export const testPluginWithOptions: FastifyPlugin<TestOptions> = fp(
  function (fastify, options, _next) {
    fastify.decorate('utility', () => options.customNumber)
  },
  '>=1'
);
```

File "src/server.ts"
```js
import * as Fastify from 'fastify';

(async () => {
  try {
    const server = Fastify.fastify();
    server.register(require('./plugin'));
    await server.listen(3000, '0.0.0.0');
  } catch (err) {
    console.error(err);
  }
})();
```

Compile TypeScript to JavaScript. For example:

`$ node_modules/.bin/tsc -p . && node build/src/server.js`

## Expected behavior

No TypeError

## Your Environment

- *Node version*: 12
- *Fastify version*: 3.0.0-rc.4
- *fastify-plugin version*: 2.0.0
- *TypeScript version*: 3.9.5
- *os*: Linux

**Comment by mcollina** [comment]
what are your tsc options? Given that the tests in this repos are passing, that’s the only explanation I can find.

However that does not seem a compile error, but rather a runtime one.

**Comment by persona0591** [comment]
Hi Matteo,

Thanks! :)

My tsconfig.json:

```json
{
  "extends": "./node_modules/gts/tsconfig-google.json",
  "compilerOptions": {
    "rootDir": ".",
    "outDir": "build"
  },
  "include": [
    "src/**/*.ts",
    "test/**/*.ts"
  ]
}
```

...where `./node_modules/gts/tsconfig-google.json` looks like [this](https://github.com/google/gts/blob/master/tsconfig-google.json).

You're right -- this is a runtime error, not a compile error. The transpiled JavaScript of `src/plugin.ts` looks like this:

```js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.testPluginWithOptions = void 0;
const fastify_plugin_1 = require("fastify-plugin");
exports.testPluginWithOptions = fastify_plugin_1.default(function (fastify, options, _next) {
    fastify.decorate('utility', () => options.customNumber);
}, '>=1');
//# sourceMappingURL=plugin.js.map
```

I'm not sure why `fastify_plugin_1.default` is used (rather than just `fastify_plugin_1`).

**Comment by mcollina** [comment]
I think you need `"esModuleInterop": true`. 
Or maybe you need to import `import Fastifyplugin from 'fastify-plugin';`.

cc @Ethan-Arrowood

**Comment by Ethan-Arrowood** [comment]
Yes please try with `"esModuleInterop": true` you should also specify the `"lib"` and `"module"` options.

I recommend always starting a tsconfig with `npx typescript --init`.

**Comment by persona0591** [comment]
Hi guys,

Thank you! `"esModuleInterop": true` indeed solves the issue - sorry I missed that one!

However, if I run my example code again, a new error is thrown:

`Error: plugin must be a function or a promise`

Full stack:

```
Error: plugin must be a function or a promise
    at assertPlugin (/usr/src/app/node_modules/avvio/boot.js:194:11)
    at Boot._addPlugin (/usr/src/app/node_modules/avvio/boot.js:226:3)
    at Boot.use (/usr/src/app/node_modules/avvio/boot.js:202:25)
    at Object.server.<computed> [as register] (/usr/src/app/node_modules/avvio/boot.js:32:14)
    at /usr/src/app/build/src/server.js:26:16
    at Object.<anonymous> (/usr/src/app/build/src/server.js:32:3)
    at Module._compile (internal/modules/cjs/loader.js:759:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:770:10)
    at Module.load (internal/modules/cjs/loader.js:628:32)
    at Function.Module._load (internal/modules/cjs/loader.js:555:12)
```

Am I overlooking something trivial again?

**Comment by mcollina** [comment]
likely. what's the output inside your js files?

**Comment by persona0591** [comment]
The JavaScript of file "src/plugin.ts":

```js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.testPluginWithOptions = void 0;
const fastify_plugin_1 = __importDefault(require("fastify-plugin"));
exports.testPluginWithOptions = fastify_plugin_1.default(function (fastify, options, _next) {
    fastify.decorate('utility', () => options.customNumber);
}, '>=1');
//# sourceMappingURL=plugin.js.map
```

And the JavaScript of file "src/server.ts":

```js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const Fastify = __importStar(require("fastify"));
(async () => {
    try {
        const server = Fastify.fastify();
        server.register(require('./plugin'));
        await server.listen(3000, '0.0.0.0');
    }
    catch (err) {
        console.error(err);
    }
})();
//# sourceMappingURL=server.js.map
```

**Comment by mcollina** [comment]
Your ts options does not generate the correct things :(. I’m not sure how to help here.

Hopefully @Ethan-Arrowood or @delvedor could be of more help.

**Comment by Ethan-Arrowood** [comment]
If you are going to use `require('./plugin')` directly inline of `register` make sure the plugin itself is the default export for the file

**Comment by persona0591** [comment]
🤦

Thanks Ethan - much appreciated!

**Comment by cmawhorter** [comment]
fwiw the only thing i could get to work in ts was `const fp = require('fastify-plugin')`.

i've been porting existing projects and i've never had any luck with esModuleInterop as it always seems to break other deps and fixing all the issues becomes a game of whack-a-mole where one fix always breaks something else.

**Comment by fox1t** [comment]
@cmawhorter may I ask you the version of fastify-plugin you are using? This should be fixed 100% in 2.3.0 and above, no matter if esModuleInterop is true or false. If you are using the last version, can you provide a repro repository?
